Avoid Earth

JoeCohen opened the pull request:
Uses the smallest Location containing lat/lng instead of Earth, for Observations.

(Explanation: Many recent Observations have lat/lng with Location Earth. Most (all? 99%) of them some from the mobile app or otherwise from the API. This frustrates mobile app users. It also stops the Observation from appearing on maps of its taxon (the location is too vague). This PR aims to improve things by preferring the smallest reasonable Location to `Earth` if an Observation has a geolocation.)

The PR adds:
- a Location scope, `with_minimum_bounding_box_containing_point`
- an Observation before_save callback, `prefer_minimum_bounding_box_to_earth`
- corresponding code in the Observation API

## Suggest Manual Tests
### Using the website
- [Create Observation](http://localhost:3000/observations/new)
  - Locality: `Earth`
  - Latitude: `46.438`
  - Longitude: `-121.84`
  - `Create`
  
**Expected result: (new) Observation with Location: Cispus Environmental Learning Center**

### Using the API
- Enter the following on the unix command line, substituting a valid API Key for `xxx`
```zsh
curl -X POST "http://localhost:3000/api2/observations?api_key=xxx&location=Earth&latitude=46.438&longitude=-121.847&date=2024-11-05"
```
- The response should be something like
```xml
<?xml version="1.0" encoding="UTF-8"?>
<response xmlns="http://localhost:3000/response.xsd">
  ...
  <results number="1">
    <result id="565807" type="observation"/>
  </results>
  ...
</response>
```
- View the Observation which has the `id` displayed in the above response.

**Expected result: Observation has Location: Cispus Environmental Learning Center**

JoeCohen committed:
Prefer smallest relevant Location to Earth

- Uses smallest existing Location containing lat/lon if user leaves Where blank (or manually sets it to Earth)

JoeCohen committed:
Scope `with_mininum_bounding_rectangle ...`

JoeCohen committed:
Make API prefer mbr to Earth

- API counterpart of Observation callback

JoeCohen committed:
Improve method names

- shorten "rectangle" to "box"
- conform API and model method names.

JoeCohen committed:
Fix bug in scope with_minimum_bounding_box_containing_point

- It was returning the calling Location::ActiveRecord_Relation if that Relation did not contain the point

nimmolo commented:
Allow "Earth" with lat/lng? Heck no. 

I don't feel we should allow Earth as an obs location, ever.

JoeCohen committed:
Fix bug caused by Obs outside the unknown location

- Uses the unknown location if Obs has geoloc but no minimum bounding box.
As of 20241105, that's possible because the live db unknown
 location does not contain the entire globe. Its boundaries:
 ```
             north: 89,
  west: -179,          east: 179,
            south: -89
```

JoeCohen committed:
Restore line inadvertently dropped in prior commit

- Fixes failures caused by prior commit

JoeCohen commented:
> Allow "Earth" with lat/lng? Heck no.
> 
> I don't feel we should allow Earth as an obs location, ever.

Thanks @nimmolo. I agree about "ever", but that's for another PR.
